docs/mergeop/2_conflict_detection.md

```diff
@@ -0,0 +1,205 @@
+* Fuse-overlayfs
+    * Pros
+        * Fully emulated vfs of overlay, including write operations
+    * Cons
+        * In C
+        * Not easily usable as a library, difficult-to-impossible to disentangle from FUSE (which we don’t need)
+        * No serder
+* Write our own
+    * Pros
+        * No issues with importing or externally caused breaking changes
+        * Can implement exactly the functionality we need now rather than pull in lots of code we don’t need
+    * Cons
+        * Have to write our own versions of code that already exists in one form or another in some external libraries
+
+Overall, gvisor’s library was the strongest third-party option, but the current iteration of this design does not actually require support for emulating write operations on an overlay vfs (just read-only stats) so gvisor’s biggest pro is not very meaningful right now.
+
+Given the significant maintenance cost of importing and using that third-party library, it was ruled out for now. If, in the future, we need to support write operations on an overlay vfs, gvisor can be re-evaluated against the cost of implementing that functionality in Buildkit alone. The change can be made at that time in a completely backwards-compatible way given that this choice is buried deep in the Buildkit daemon’s implementation and not a direct influence on any public interfaces.
+
+## Serialization Alternatives
+
+There were several choices for which format to use when (de)serializing layer filesystem metadata:
+
+* Tar
+    * Pros
+        * Already in heavy use in the codebase, plenty of util functions to either re-use or use as a reference
+        * Supports all the file metadata we need while also allowing us to skip storing actual file contents
+    * Cons
+        * The format is not perfectly optimized for our use case, so there may be more overhead than strictly necessary when converting between its format and our in-memory data structures used for conflict detection
+        * The tar format is not exactly a pleasure to use as a developer (though this is partially alleviated by the existing tooling around it)
+* [Continuity’s Manifest+Resource Proto](https://github.com/containerd/continuity/blob/33a84563724b68f70fc71fbb632cf9e0ad60ba5d/proto/manifest.proto#L7)
+    * Pros
+        * An existing format that has already underwent careful thought+design in terms of creating platform-agnostic representations of filesystems
+        * Proto is already used in the codebase for serder of API objects
+    * Cons
+        * The format is not perfectly optimized for our use case of conflict detection (same as Tar)
+        * Not clear if the continuity package is maintained much these days besides the util functions under fs used by containerd’s differ (which the proto is separate from). We may become partially or fully responsible for maintenance of the proto and thus have to deal with any other consumers of it that exist out in the world.
+* Custom Proto or JSON
```

Comment:
[TOC by stargz/eStargz](https://github.com/containerd/stargz-snapshotter/blob/170bca44f2862599cbc209a8716ec3d008bf403c/estargz/types.go#L93-L186) ([doc](https://github.com/containerd/stargz-snapshotter/blob/v0.6.4/docs/stargz-estargz.md#toc-tocentries-and-footer)) might be another good candidate of JSON-formatted filesystem metadata :)
(zstd:chunked by Podman also uses the similar one.)

Reply:
Ah interesting, thanks for pointing me to that! I never thought to look but makes sense that stargz had to solve a similar problem too. I think for now I still want to go with the proposal in the doc of using tar because it simplifies some of details around import/export of this metadata and doesn't tie us into any changes you may need to make to the stargz format in the future.

However, I added a section mentioning it in case we ever want to reconsider this decision in the future and are looking back for other options.